pkg/optimizer/datapath_csp.go

```diff
@@ -0,0 +1,700 @@
+// The trick is to use the dot-product constraint, add the summing var with weight -1 and force the result to be 0
+func (dpc *DataPathCSP) setVarAsWeightedSum(sumVarname string, arrayToSum, weights []string) {
+	arrayToSum = append(arrayToSum, sumVarname)
+	weights = append(weights, "-1")
+	dpc.fzModel.AddConstraint(
+		IntLinEqConstraint,
+		[]string{fznCompoundLiteral(weights, false), fznCompoundLiteral(arrayToSum, false), "0"},
+		GetDefinesVarAnnotation(sumVarname),
+	)
+}
+
+// Returns which actions should be activated by the module at position pathPos (according to the solver's solution)
+func (dpc *DataPathCSP) getSolutionActionsAtPos(solverSolution CPSolution, pathPos int) []taxonomy.Action {
+	actions := []taxonomy.Action{}
+	for _, action := range dpc.problemData.Actions {
+		actionVarname := getActionVarname(action)
+		actionSolution := solverSolution[actionVarname]
+		if actionSolution[pathPos] == TrueValue {
+			actions = append(actions, action)
+		}
+	}
+	return actions
+}
+
+// Translates a solver's solution into a FybrikApplication Solution for a given data-path
+// TODO: better handle error messages
+func (dpc *DataPathCSP) decodeSolverSolution(solverSolutionStr string, pathLen int) (Solution, error) {
+	solverSolution, err := dpc.fzModel.ReadBestSolution(solverSolutionStr)
+	if err != nil {
+		return Solution{}, err
+	}
+	if len(solverSolution) == 0 {
+		return Solution{}, nil // UNSAT
+	}
+
+	modCapSolution := solverSolution[modCapVarname]
+	clusterSolution := solverSolution[clusterVarname]
+	saSolution := solverSolution[saVarname]
+	srcIntfcSolution := solverSolution[srcIntfcVarname]
+	sinkIntfcSolution := solverSolution[sinkIntfcVarname]
+
+	srcIntfcIdx, _ := strconv.Atoi(srcIntfcSolution[0])
+	srcNode := &Node{Connection: dpc.reverseIntfcMap[srcIntfcIdx]}
+
+	solution := Solution{}
+	for pathPos := 0; pathPos < pathLen; pathPos++ {
+		modCapIdx, _ := strconv.Atoi(modCapSolution[pathPos])
+		modCap := dpc.modulesCapabilities[modCapIdx-1]
+		clusterIdx, _ := strconv.Atoi(clusterSolution[pathPos])
+		saIdx, _ := strconv.Atoi(saSolution[pathPos])
+		sa := appApi.FybrikStorageAccountSpec{}
+		if saIdx > 0 { // recall that a value of 0 means no storage account
+			sa = dpc.env.StorageAccounts[saIdx-1].Spec
+		}
+		sinkIntfcIdx, _ := strconv.Atoi(sinkIntfcSolution[pathPos])
+		sinkNode := &Node{Connection: dpc.reverseIntfcMap[sinkIntfcIdx], Virtual: modCap.virtualSink}
+		edge := Edge{Module: modCap.module, CapabilityIndex: modCap.capabilityIdx, Source: srcNode, Sink: sinkNode}
+		resolvedEdge := ResolvedEdge{
+			Edge:           edge,
+			Actions:        dpc.getSolutionActionsAtPos(solverSolution, pathPos),
+			Cluster:        dpc.env.Clusters[clusterIdx-1].Name,
+			StorageAccount: sa,
+		}
+		solution.DataPath = append(solution.DataPath, &resolvedEdge)
+		srcNode = sinkNode
+	}
+
+	if dpc.problemData.Context.Flow == taxonomy.WriteFlow { // reverse solution
```

Comment:
I believe this same functionality is implemented in the non-csp path builder.  Perhaps create a common function?

Reply:
Common code moved to `solution.go`